system-configuration/src/network_configuration.rs

```diff
@@ -0,0 +1,452 @@
+            state_domain_name: domain_name.0,
+            setup_domain_name: domain_name.1,
+            state_server_addresses: server_addresses.0,
+            setup_server_addresses: server_addresses.1,
+        }
+    }
+    
+    /// Returns DomainName (state and setup)
+    pub fn domain_name(&self) -> (Option<String>, Option<String>) {
+        (
+            self.state_domain_name.clone(),
+            self.setup_domain_name.clone(),
+        )
+    }
+
+    /// Returns ServerAddresses (state and setup)
+    pub fn server_addresses(&self) -> (Option<Vec<IpAddr>>, Option<Vec<IpAddr>>) {
+        (
+            self.state_server_addresses.clone(),
+            self.setup_server_addresses.clone(),
+        )
+    }
+}
+
+/// Global network object
+pub struct SCNetworkGlobal;
+
+impl SCNetworkGlobal {
+    fn query(session_name: &str, key: &str) -> Option<String> {
+        let store = SCDynamicStoreBuilder::new(session_name).build();
+        let path = CFString::from_static_string("State:/Network/Global/IPv4");
+
+        if let Some(value) = store.get(path.clone()) {
+            if let Some(dict) = value.downcast_into::<CFDictionary>() {
+                if let Some(val) = dict.find2(&CFString::new(key)) {
+                    let value = unsafe { CFType::wrap_under_get_rule(val) };
+                    if let Some(value) = value.downcast::<CFString>() {
+                        return Some(value.to_string());
+                    }
+                }
+            }
+        }
+
+        return None;
+    }
+
+    /// Returns primary network service
+    pub fn service(&self) -> Option<SCNetworkService> {
+        if let Some(service_id) = SCNetworkGlobal::query("ng_service", "PrimaryService") {
+            for _service in SCNetworkService::list() {
+                if _service.id() == service_id {
+                    return Some(_service);
+                }
+            }
+        }
+
+        return None;
+    }
+
+    /// Returns primary network interface
+    pub fn interface(&self) -> Option<SCNetworkInterface> {
+        if let Some(ifname) = SCNetworkGlobal::query("ng_interface", "PrimaryInterface") {
+            for iface in SCNetworkInterface::list() {
+                if let Some(bsd_name) = iface.bsd_name() {
+                    if bsd_name == ifname {
+                        return Some(iface);
+                    }
+                }
+            }
+        }
+
+        return None;
+    }
+
+    /// Returns default route on primary network service.
+    pub fn router(&self) -> Option<IpAddr> {
+        if let Some(router_str) = SCNetworkGlobal::query("ng_interface_router", "Router") {
+            if let Ok(router_ip) = router_str.parse::<IpAddr>() {
+                return Some(router_ip);
+            }
+        }
+
+        return None;
+    }
+
+    // pub fn netinfo(&self);
+    // pub fn proxies(&self) ;
+}
+
+/// Network service object.
+pub struct SCNetworkService(pub SCNetworkServiceRef);
```

Comment:
I think all of these types should be declared according to the standard used in `core-foundation`, with the macros there. Compare with how I have done it for `SCDynamicStore`: https://github.com/mullvad/system-configuration-rs/blob/master/system-configuration/src/dynamic_store.rs#L137

That way we'll get automatic and correct `Drop` implementation and lots of helper methods.

Reply:
Okay.